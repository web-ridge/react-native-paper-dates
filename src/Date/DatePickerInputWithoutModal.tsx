import TextInputWithMask from '../TextInputMask'
import { HelperText, useTheme } from 'react-native-paper'
import { StyleSheet, View } from 'react-native'
import useDateInput from './inputUtils'
import type { DatePickerInputProps } from './DatePickerInput.shared'
import { forwardRef, ReactNode } from 'react'

function DatePickerInputWithoutModal(
  {
    label,
    value,
    onChange,
    style,
    locale,
    validRange,
    inputMode,
    withDateFormatInLabel = true,
    hasError,
    hideValidationErrors,
    onValidationError,
    modal,
    inputButton,
    saveLabel,
    saveLabelDisabled,
    uppercase,
    startYear,
    endYear,
    monthOnly = false,
    onChangeText,
    inputEnabled,
    disableStatusBarPadding,
    startWeekOnMonday,
    ...rest
  }: DatePickerInputProps & {
    modal?: (params: {
      value: DatePickerInputProps['value']
      locale: DatePickerInputProps['locale']
      inputMode: DatePickerInputProps['inputMode']
      validRange: DatePickerInputProps['validRange']
      saveLabel: DatePickerInputProps['saveLabel']
      saveLabelDisabled: DatePickerInputProps['saveLabelDisabled']
      uppercase: DatePickerInputProps['uppercase']
      startYear: DatePickerInputProps['startYear']
      endYear: DatePickerInputProps['endYear']
      monthOnly: DatePickerInputProps['monthOnly']
      inputEnabled: DatePickerInputProps['inputEnabled']
      disableStatusBarPadding: DatePickerInputProps['disableStatusBarPadding']
      startWeekOnMonday?: DatePickerInputProps['startWeekOnMonday']
    }) => any
    inputButton?: ReactNode
  },
  ref: any
) {
  const theme = useTheme()

  const {
    formattedValue,
    inputFormat,
    onChangeText: onDateInputChangeText,
    error,
  } = useDateInput({
    locale,
    value,
    validRange,
    inputMode,
    monthOnly,
    onChange,
    onValidationError,
  })

  let disabled

  if (inputEnabled !== undefined) {
    disabled = !inputEnabled
  }

  if (rest.disabled) {
    disabled = rest.disabled
  }

  return (
    <>
      <View style={styles.root}>
        <View style={styles.inputContainer}>
          <TextInputWithMask
            {...rest}
            ref={ref}
            label={getLabel({
              // TODO: support label components?
              label: label as any,
              inputFormat,
              withDateFormatInLabel,
            })}
            value={formattedValue}
            keyboardType={rest.keyboardType ?? 'number-pad'}
            mask={inputFormat}
            disabled={disabled}
            onChangeText={onDateInputChangeText}
            onChange={(e) => onChangeText && onChangeText(e.nativeEvent.text)}
            keyboardAppearance={
              rest.keyboardAppearance ?? (theme.dark ? 'dark' : 'default')
            }
            error={(!!error && !hideValidationErrors) || !!hasError}
            style={[styles.input, style]}
            inputButton={inputButton}
          />
        </View>
        {error && !hideValidationErrors ? (
          <HelperText type="error" visible={!!error}>
            {error}
          </HelperText>
        ) : null}
      </View>
      {modal?.({
        value,
        locale,
        inputMode,
        validRange,
        saveLabel,
        saveLabelDisabled,
        uppercase,
        startYear,
        endYear,
        monthOnly,
        inputEnabled,
        disableStatusBarPadding,
        startWeekOnMonday,
      })}
    </>
  )
}

function getLabel({
  withDateFormatInLabel,
  inputFormat,
  label,
}: {
  withDateFormatInLabel: boolean
  inputFormat: string
  label: string | undefined
}) {
  if (withDateFormatInLabel) {
    return label ? `${label} (${inputFormat})` : inputFormat
  }
  return label || ''
}

const styles = StyleSheet.create({
  root: {
    flex: 1,
    flexGrow: 1,
    justifyContent: 'center',
    alignItems: 'flex-start',
    width: '100%',
  },
  inputContainer: {
    flexGrow: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    width: '100%',
  },
  input: {
    flexGrow: 1,
    width: '100%',
  },
})
export default forwardRef(DatePickerInputWithoutModal)
