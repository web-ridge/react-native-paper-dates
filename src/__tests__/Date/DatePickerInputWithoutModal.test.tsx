import { render } from '@testing-library/react-native'
import DatePickerInputWithoutModal from '../../Date/DatePickerInputWithoutModal'

// 7th of August, 2025
const fixedDate = new Date(2025, 7, 7)

it('renders DatePickerInputWithoutModal', () => {
  const { toJSON } = render(
    <DatePickerInputWithoutModal
      locale="en"
      value={fixedDate}
      onChange={() => null}
      inputMode="start"
    />
  )
  expect(toJSON()).toMatchSnapshot()
})
