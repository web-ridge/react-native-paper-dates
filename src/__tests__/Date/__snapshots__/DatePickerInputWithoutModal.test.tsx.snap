// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DatePickerInputWithoutModal 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
      "flex": 1,
      "flexGrow": 1,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 1,
        "justifyContent": "flex-start",
        "width": "100%",
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          {
            "flexGrow": 1,
            "width": "100%",
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          [
            {
              "bottom": 0,
              "height": 2,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "zIndex": 1,
            },
            {
              "height": 1,
            },
            {
              "backgroundColor": "rgba(73, 69, 79, 1)",
              "transform": [
                {
                  "scaleY": 0.5,
                },
              ],
            },
            undefined,
          ]
        }
        testID="text-input-underline"
      />
      <View
        onLayout={[Function]}
        style={
          [
            {
              "flexGrow": 1,
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "minHeight": 56,
            },
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            [
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "overflow": "hidden",
              },
              {
                "zIndex": 3,
              },
            ]
          }
        >
          <View
            collapsable={false}
            pointerEvents="none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": [
                  {
                    "translateX": 4,
                  },
                ],
                "width": 750,
              }
            }
          >
            <View
              style={
                {
                  "width": 97.33333333333333,
                }
              }
            >
              <Text
                collapsable={false}
                maxFontSizeMultiplier={1.5}
                numberOfLines={1}
                onLayout={[Function]}
                onTextLayout={[Function]}
                style={
                  {
                    "color": "rgba(103, 80, 164, 1)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "maxWidth": 97.33333333333333,
                    "opacity": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 30,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateY": -12,
                      },
                      {
                        "scale": 0.75,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
                testID="text-input-flat-label-active"
              >
                MM/DD/YYYY
              </Text>
              <Text
                collapsable={false}
                maxFontSizeMultiplier={1.5}
                numberOfLines={1}
                style={
                  {
                    "color": "rgba(73, 69, 79, 1)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "maxWidth": 97.33333333333333,
                    "opacity": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 30,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateY": -12,
                      },
                      {
                        "scale": 0.75,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
                testID="text-input-flat-label-inactive"
              >
                MM/DD/YYYY
              </Text>
            </View>
          </View>
        </View>
        <TextInput
          cursorColor="rgba(103, 80, 164, 1)"
          editable={true}
          keyboardAppearance="default"
          keyboardType="number-pad"
          maxFontSizeMultiplier={1.5}
          maxLength={10}
          multiline={false}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholderTextColor="transparent"
          selectionColor="rgba(103, 80, 164, 1)"
          style={
            [
              {
                "flexGrow": 1,
                "margin": 0,
              },
              {},
              {
                "paddingBottom": 4,
                "paddingTop": 24,
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "minWidth": 65,
                "paddingLeft": 16,
                "paddingRight": 16,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              undefined,
              [
                {},
              ],
              undefined,
            ]
          }
          testID="text-input-flat"
          underlineColorAndroid="transparent"
          value="06/06/2025"
        />
      </View>
    </View>
  </View>
</View>
`;
